// @wyw-in-js/shared/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueType = exports.slugify = exports.isFeatureEnabled = exports.logger = exports.enableDebug = exports.isBoxedPrimitive = exports.findPackageJSON = exports.hasEvalMeta = exports.syncResolve = exports.asyncResolveFallback = void 0;
var asyncResolveFallback_1 = require("./asyncResolveFallback");
Object.defineProperty(exports, "asyncResolveFallback", { enumerable: true, get: function () { return asyncResolveFallback_1.asyncResolveFallback; } });
Object.defineProperty(exports, "syncResolve", { enumerable: true, get: function () { return asyncResolveFallback_1.syncResolve; } });
var hasEvalMeta_1 = require("./hasEvalMeta");
Object.defineProperty(exports, "hasEvalMeta", { enumerable: true, get: function () { return hasEvalMeta_1.hasEvalMeta; } });
var findPackageJSON_1 = require("./findPackageJSON");
Object.defineProperty(exports, "findPackageJSON", { enumerable: true, get: function () { return findPackageJSON_1.findPackageJSON; } });
var isBoxedPrimitive_1 = require("./isBoxedPrimitive");
Object.defineProperty(exports, "isBoxedPrimitive", { enumerable: true, get: function () { return isBoxedPrimitive_1.isBoxedPrimitive; } });
var logger_1 = require("./logger");
Object.defineProperty(exports, "enableDebug", { enumerable: true, get: function () { return logger_1.enableDebug; } });
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
var isFeatureEnabled_1 = require("./options/isFeatureEnabled");
Object.defineProperty(exports, "isFeatureEnabled", { enumerable: true, get: function () { return isFeatureEnabled_1.isFeatureEnabled; } });
var slugify_1 = require("./slugify");
Object.defineProperty(exports, "slugify", { enumerable: true, get: function () { return slugify_1.slugify; } });
var types_1 = require("./types");
Object.defineProperty(exports, "ValueType", { enumerable: true, get: function () { return types_1.ValueType; } });
